models/tests/test_document.py

```diff
@@ -0,0 +1,168 @@
+
+    def test_add_incorrect_type(self):
+        with self.assertRaises(ReferenceSet.MultipleTypeError):
+            self.document.children.add(self.DemoNestedDocument(name='d'))
+
+    def test_remove(self):
+        self.document.children.remove(self.DemoDocument(name='a'))
+        self.assertFalse(self.DemoDocument(name='a') in self.document.children)
+
+    def test_delete_referee(self):
+        referee = next(iter(self.document.children))
+        referee.delete()
+        self.assertEqual(len(self.document.children), 2)
+
+    def test_find_referee(self):
+        children = [
+            self.DemoNestedDocument(name='child_1', children=[]),
+            self.DemoNestedDocument(name='child_2', children=[]),
+            self.DemoNestedDocument(name='child_3', children=[])
+        ]
+        document = self.DemoNestedDocument(name='test', children=children)
+        self.assertEqual(document.children.get('child_1'), children[0])
+
+    def test_delete_referee_by_key(self):
+        children = [
+            self.DemoNestedDocument(name='child_1', children=[]),
+            self.DemoNestedDocument(name='child_2', children=[]),
+            self.DemoNestedDocument(name='child_3', children=[])
+        ]
+        document = self.DemoNestedDocument(name='test', children=children)
+        self.assertEqual(document.children.get('child_1'), children[0])
+        del document.children['child_1']
+        self.assertIsNone(document.children.get('child_1'))
+
+    def test_delete_multiple_references(self):
+        referee = self.DemoDocument(name='e')
+        document_1 = self.DemoNestedDocument(name='test_1', children=[referee])
+        document_2 = self.DemoNestedDocument(name='test_2', children=[referee])
+        self.assertTrue(referee in document_1.children)
+        self.assertTrue(referee in document_2.children)
+        referee.delete()
+        self.assertFalse(referee in document_1.children)
+        self.assertFalse(referee in document_2.children)
+
+    def test_delete_indexed_document_reference(self):
+        referee = self.DemoNestedDocument(name='test_1', children=[])
+        document = self.DemoNestedDocument(name='test_2', children=[referee])
+        self.assertTrue(referee in document.children)
+        referee.delete()
+        self.DemoNestedDocument.reload()
+        self.assertIsNone(self.DemoNestedDocument.find('test_1'))
+        self.assertFalse(referee in document.children)
+
+    def test_delete_deep_reference(self):
+        root_referee = self.DemoDocument(name='f')
+        mid_referee = self.DemoNestedDocument(name='test_1', children=[root_referee])
+        document = self.DemoNestedDocument(name='test_2', children=[mid_referee])
+        root_referee.delete()
+        self.DemoNestedDocument.reload()
+        for referee in document.children:
+            self.assertFalse(root_referee in referee.children)
+
+    def test_typed_reference_field(self):
```

Comment:
Not sure what is being tested here, can you clarify please?

Reply:
docstring added